### ISSUE TITLE
TOC Drag&Drop: item not moved but copied

### ISSUE THREAD

**Issue by xgadscu** [kookma/TW-TOC#2]
Hi @kookma, I really like your tiddlywiki plugins, especially the TW-TOC Multi level TOC UI macro.

Unfortunately drag&drop of an item from one layer into another layer does not work. The item is not moved but copied.

I already tried to find the reason for that behaviour and put some action-logs into your code. Obviously the toc-item-dragstart-Macro is not triggered by startactions. OLDTAG is empty.

I use TiddlyWiki via a docker container. But you can find the error also in your official Tutorial.

Do you have an idea?

Best Regards
xgadscu

**Comment by kookma** [comment]
@xgadscu 
> Unfortunately drag&drop of an item from one layer into another layer does not work. 

I will investigate when I have free time. by the way I welcome your solution here!

**Comment by xgadscu** [comment]
Hi @kookma, thank you for responding.

I read your discussion with eric shulman (https://groups.google.com/g/tiddlywiki/c/_IojZZU4O0U). I think that the solutions are very similar and that there is only a very small error here. 
Another hint is that in the macro toc-item-dragtip the old tiddler (actionTiddler) is not shown. This is correct in Eric's case.

![grafik](https://user-images.githubusercontent.com/42093828/161541609-690ba8fb-4962-4238-b9ca-41f6907eb721.png)

I'm not that familiar with javascript and stylesheets. Therefore I can only give ideas here.

xgadscu

**Comment by kookma** [comment]
@xgadscu 
One difference here is the Eric solution uses tiddlers with one tag! In other words if you reorder tiddler in TOC tree by drag and drop, it is enough to remove the old tag and add the new one (where you drop the tiddler), but when you have multiple tags, the situation is complicated.

The TOC plugin does not uses any JS, so this is a limitation. I keep this issue open to see how we can resolve this.

**Comment by xgadscu** [comment]
Hi, I can imagine that multiple tags are more complicated. But as far as I investigated the problem the reason here is that **toc-item-dragstart** is not executed.

From my understanding in authormode the class tc-droppable-placeholder is activated which defines startactions for D&D. 
See $:/plugins/kookma/toc/macros/multilevel-exapndable:

```
\define toc-show-item-authormode(item, is-branch:no)
<$vars item=<<__item__>> is-branch=<<__is-branch__>> >
<$droppable actions=<<toc-item-drop>>>
	<div class="tc-droppable-placeholder kk-toc-placeholder"></div> 
	<$draggable tiddler=<<item>> startactions=<<toc-item-dragstart>> endactions=<<toc-item-dragend>> dragimagetype="blank">
	{{$item$||$(tidTocTemplate)$}}
	</$draggable>
	<div class="tc-droppable-placeholder kk-toc-placeholder"><<toc-item-dragtip>></div> 
</$droppable>
</$vars>
\end

\define toc-show-item-readermode(item, is-branch:no)
<$vars item=<<__item__>> is-branch=<<__is-branch__>> >
    {{$item$||$(tidTocTemplate)$}}
</$vars>
\end
```
In this case toc-item-dragstart should be triggered. 
I put some action-log-statements into the code of $:/plugins/kookma/toc/macros/multilevel-draggable:

```
\define toc-item-dragdata()      $:/temp/toc/dragdata
\define toc-item-dragdata-path() $:/temp/toc/dragdata/path
\define toc-item-dragstart()   
<$action-log $$message="Log from toc-item-dragstart: Start" $$filter="actionTiddler"/>
<$action-setfield  $tiddler=<<toc-item-dragdata>> text=<<actionTiddler>> />
<$action-setfield  $tiddler=<<toc-item-dragdata-path>> text=<<path>> />
<$action-log $$message="Log from toc-item-dragstart" $$filter="toc-item-dragdata toc-item-dragdata-path"/>
\end
\define toc-item-dragend()  
<$action-log $$message="Log from toc-item-dragend" $$filter="actionTiddler"/>
<$action-deletetiddler $tiddler=<<toc-item-dragdata>> />
<$action-deletetiddler $tiddler=<<toc-item-dragdata-path>> />
\end

\define toc_caption()
<$transclude field="caption">
   <$view field="title"/>
</$transclude>
\end

\define toc-item-dragtip()
<$wikify name=actionTiddler text={{{ [<toc-item-dragdata>get[text]] }}}><!--handles title with space in title-->
<$vars actionTiddlerx={{{ [<toc-item-dragdata>get[text]] }}}>
<$list filter="[<actionTiddler>!match<item>]"> <!-- DON'T SHOW TIP ON SELF -->
   <div class="kk-toc-dragtip">
      <nobr>drop <$tiddler tiddler=<<actionTiddlerx>>>''<<toc_caption>>''</$tiddler></nobr><br>
      <nobr>onto <$tiddler tiddler=<<item>>         >''<<toc_caption>>''</$tiddler></nobr><br>
      <div style="font-size:80%; border-top:1px solid; margin-top:0.2em;">
         NO KEY=insert before, SHIFT=insert after, CTRL=drop into
      </div>
   </div>
</$list>
</$vars>
</$wikify>
\end


\define toc-item-drop()
<$reveal default=<<actionTiddler>> type="nomatch" text=<<item>>> <!-- DON'T DROP ON SELF -->
   <!-- get OLDTAG and NEWTAG, exclude special TiddlyBook tags -->
   <$vars sourcepath={{{ [<toc-item-dragdata-path>get[text]] }}}
          targetpath=<<path>> >
   <$vars oldtag= {{{ [<sourcepath>split[/]]-[<actionTiddler>]+[last[]]  }}}
           newtag={{{ [<targetpath>split[/]]-[<item>]+[last[]]  }}}>
<$action-log $$message="Log from toc-item-drop" $$filter="actionTiddler targetpath sourcepath oldtag newtag"/>

   <$reveal default=<<modifier>> type="match" text="normal"> <!-- DROP BEFORE SIBLING -->
      <$set name="order" filter="[<newtag>tagging[]]">
      <$list filter=<<order>>> <$action-deletefield $field="list-before"/> <$action-deletefield $field="list-after"/> </$list>
      <$action-listops $tiddler=<<actionTiddler>> $field="tags" $subfilter="-[<oldtag>] [<newtag>]" />
      <$action-listops $tiddler=<<oldtag>>        $field="list" $subfilter="-[<actionTiddler>]" />
      <$action-listops $tiddler=<<newtag>>        $field="list" $subfilter="[enlist<order>] -[<actionTiddler>] [<actionTiddler>] +[putbefore<item>]" />
      </$set>
   </$reveal>
   <$reveal default=<<modifier>> type="match" text="shift"> <!-- DROP AFTER SIBLING -->
      <$set name="order" filter="[<newtag>tagging[]]">
      <$list filter=<<order>>> <$action-deletefield $field="list-before"/> <$action-deletefield $field="list-after"/> </$list>
      <$action-listops $tiddler=<<actionTiddler>> $field="tags" $subfilter="-[<oldtag>] [<newtag>]" />
      <$action-listops $tiddler=<<oldtag>>        $field="list" $subfilter="-[<actionTiddler>]" />
      <$action-listops $tiddler=<<newtag>>        $field="list" $subfilter="[enlist<order>] -[<actionTiddler>] [<actionTiddler>] +[putafter<item>]" />
      </$set>
   </$reveal>
   <$reveal default=<<modifier>> type="match" text="ctrl"> <!-- DROP AS CHILD -->
      <$set name="order" filter="[<item>tagging[]]">
      <$list filter=<<order>>> <$action-deletefield $field="list-before"/> <$action-deletefield $field="list-after"/> </$list>
      <$action-listops $tiddler=<<actionTiddler>> $field="tags" $subfilter="-[<oldtag>] [<item>]" />
      <$action-listops $tiddler=<<oldtag>>        $field="list" $subfilter="-[<actionTiddler>]" />
      <$action-listops $tiddler=<<item>>          $field="list" $subfilter="[enlist<order>] -[<actionTiddler>] [<actionTiddler>]" />
      </$set>
   </$reveal>
   </$vars>
   </$vars>
</$reveal>
\end
```

In the log you can see only "Log from toc-item-drop". Neither the log-item "Log from toc-item-dragstart" nor "Log from toc-item-drag-end" does appear in the log:

![grafik](https://user-images.githubusercontent.com/42093828/162401916-a18e6b7c-b976-4646-b533-10b52828d28d.png)

Obviously the macro toc-item-dragstart and as a consequence the macros **toc-item-dragdata** and **toc-item-dragdata-path** are not called. sourcepath and oldtag are empty strings and so oldtag cannot be deleted:

```
      <$action-listops $tiddler=<<actionTiddler>> $field="tags" $subfilter="-[<oldtag>] [<newtag>]" />
      <$action-listops $tiddler=<<oldtag>>        $field="list" $subfilter="-[<actionTiddler>]" />

```

I may be misunderstanding the D&D mechanism because I'm not the expert here. But I hope my observations can help you to find the reason for the problem.

Thank you for your time
xgadscu

**Comment by kookma** [comment]
Thank you for the debug info. I am working on the issue. One solution is to restrict the drag and drop to item with single tag as you can see in Eric code!

**Comment by kookma** [comment]
This issue is fixed in new release 1.4.1
